sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -130,8 +130,24 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           if limit < conf.topKSortFallbackThreshold =>
         Some(TakeOrderedAndProjectExec(
           limit, order, projectList, planLater(child)))
+      case Sort(order, true, child) if supportTakeOrdered(child) =>
+        Some(TakeOrderedAndProjectExec(
+          child.maxRows.get.toInt, order, child.output, planLater(child)))
+      case Project(projectList, Sort(order, true, child)) if supportTakeOrdered(child) =>
+        Some(TakeOrderedAndProjectExec(
+          child.maxRows.get.toInt, order, projectList, planLater(child)))
       case _ => None
     }
+
+    private def supportTakeOrdered(plan: LogicalPlan): Boolean = {
+      // Maximum support up to 655360, a larger number of rows does not improve performance.
+      plan.maxRows.exists(_ < math.min(conf.topKSortFallbackThreshold, 655360)) &&
```

Comment:
shall we add a new config for this optimization?

Reply:
OK. new config: `spark.sql.execution.topKSortMaxRowsThreshold`.